packages/core/src/tools/ContactProvider.ts

```diff
@@ -0,0 +1,171 @@
+  private _getReadContactPermission = (): Permission => {
+    switch (Platform.OS) {
+      case 'ios':
+        return PERMISSIONS.IOS.CONTACTS;
+      case 'android':
+        return PERMISSIONS.ANDROID.READ_CONTACTS;
+      default:
+        throw new Error('Unsupported device.');
+    }
+  };
+
+  private _showToast = (type: string, message: string): void => {
+    Toast.show({
+      type,
+      position: 'bottom',
+      bottomOffset: 20,
+      text1: type === 'success' ? 'Success' : 'Error',
+      text2: message,
+    });
+  };
+
+  private _requestReadContactsPermission =
+    async (): Promise<PermissionResult> => {
+      try {
+        const reactContactPermission = this._getReadContactPermission();
+        const permissionStatus = await request(reactContactPermission);
+        return permissionStatus as PermissionResult;
+      } catch (error) {
+        console.error('Error requesting contacts permission:', error);
+        return PermissionResult.DENIED;
+      }
+    };
+
+  private _createContactData = (
+    contactData: ContactData,
+  ): Partial<Contacts.Contact> => {
+    const {firstName, lastName, email, phoneNumbers, address, notes} =
+      contactData;
+
+    const contact: Partial<Contacts.Contact> = {
+      givenName: firstName,
+      familyName: lastName || undefined,
+      emailAddresses: email
+        ? [
+            {
+              label: 'email',
+              email,
+            },
+          ]
+        : undefined,
+      phoneNumbers: phoneNumbers
+        ? [
+            {
+              label: 'mobile',
+              number: phoneNumbers.mobilePhone,
+            },
+            {
+              label: 'fixed',
+              number: phoneNumbers.fixedPhone,
+            },
+          ]
+        : undefined,
+      postalAddresses: address
+        ? [
+            {
+              label: 'main address',
+              formattedAddress: address,
+              street: null,
+              pobox: null,
+              neighborhood: null,
+              city: null,
+              region: null,
+              state: null,
+              postCode: null,
+              country: null,
+            },
+          ]
+        : undefined,
+      note: notes || undefined,
+    };
+
+    return contact;
+  };
+
+  saveContact = async (contactData: ContactData): Promise<boolean> => {
+    try {
+      if (!contactData.firstName) {
+        throw new Error('Required contact data is missing.');
+      }
+
+      const permissionResult = await this._requestReadContactsPermission();
+
+      let toastType: string;
+      let toastMessage: string;
+
+      switch (permissionResult) {
+        case PermissionResult.GRANTED:
+        case PermissionResult.LIMITED:
+          const _contactData = this._createContactData(contactData);
+          await Contacts.addContact(_contactData);
+          toastType = 'success';
+          toastMessage = `Contact ${contactData.firstName} added successfully.`;
+          break;
```

Comment:
Even when granting access to contacts when asked, I got an error while creating contact : "java.lang.SecurityException: Permission denied writing com.android.provider.contacts.Contact..."